### PR TITLE
Fix import date

### PR DESCRIPTION
**Pull request by archetipo**

### CODE REVIEW

File: l10n_it_fatturapa_in/wizard/wizard_import_fatturapa.py [hunk 1]
```diff
@@ -1274,6 +1299,15 @@ def importFatturaPA(self, cr, uid, ids, context=None):
...
             xml_string = self.remove_xades_sign(xml_string)
             xml_string = self.strip_xml_content(xml_string)
+            # Fix Date format
+            # Date format must be YYYY-MM-DD
+            # DateType in XSD implements
+            # restrict xs:date minInclusive YYYY-MM-DD  and date should be
+            # in others not parsable ISO 8601:2004 formats like:
+            # YYYY-MM-DDT00:00:00+00:00
+            # YYYY-MM-DD+00:00
+            # YYYY-MM-DDT00:00:00Z
```
Comment (eLBati) [inline l10n_it_fatturapa_in/wizard/wizard_import_fatturapa.py]: Caricando il file IT05979361218_007 su http://sdi.fatturapa.gov.it/SdI2FatturaPAWeb/AccediAlServizioAction.do?pagina=controlla_fattura ottengo

```
File not compliant with the format
Check riga: 85 - colonna: 89  '2015-06-03T10:21:58+00:00' is not a valid value for 'date'.  Check that the file transmitted complies with the format expected   00200
File not compliant with the format
Check riga: 85 - colonna: 89  The value '2015-06-03T10:21:58+00:00' of element 'DataScadenzaPagamento' is not valid.    Check that the file transmitted complies with the format expected   00200
File not compliant with the format
Check riga: 91 - colonna: 77  '2015-06-04+1:00' is not a valid value for 'date'.    Check that the file transmitted complies with the format expected   00200
File not compliant with the format
Check riga: 91 - colonna: 77  The value '2015-06-04+1:00' of element 'DataDecorrenzaPenale' is not valid.   Check that the file transmitted complies with the format expected   00200
```

Quindi direi che il file non è da accettare

Comment (archetipo) [inline l10n_it_fatturapa_in/wizard/wizard_import_fatturapa.py]: grazie verifico..

2015-06-05 9:41 GMT+02:00 Lorenzo Battistini notifications@github.com:

> In l10n_it_fatturapa_in/wizard/wizard_import_fatturapa.py
> https://github.com/eLBati/l10n-italy/pull/61#discussion_r31793002:
> 
> > @@ -1274,6 +1299,15 @@ def importFatturaPA(self, cr, uid, ids, context=None):
> >                  xml_string = fatturapa_attachment.datas.decode('base64')
> >              xml_string = self.remove_xades_sign(xml_string)
> >              xml_string = self.strip_xml_content(xml_string)
> > -            # Fix Date format
> > -            # Date format must be YYYY-MM-DD
> > -            # DateType in XSD implements
> > -            # restrict xs:date minInclusive YYYY-MM-DD  and date should be
> > -            # in others not parsable ISO 8601:2004 formats like:
> > -            # YYYY-MM-DDT00:00:00+00:00
> > -            # YYYY-MM-DD+00:00
> > -            # YYYY-MM-DDT00:00:00Z
> 
> Caricando il file IT05979361218_007 su
> http://sdi.fatturapa.gov.it/SdI2FatturaPAWeb/AccediAlServizioAction.do?pagina=controlla_fattura
> ottengo
> 
> File not compliant with the format
> Check riga: 85 - colonna: 89  '2015-06-03T10:21:58+00:00' is not a valid value for 'date'.  Check that the file transmitted complies with the format expected   00200
> File not compliant with the format
> Check riga: 85 - colonna: 89  The value '2015-06-03T10:21:58+00:00' of element 'DataScadenzaPagamento' is not valid.    Check that the file transmitted complies with the format expected   00200
> File not compliant with the format
> Check riga: 91 - colonna: 77  '2015-06-04+1:00' is not a valid value for 'date'.    Check that the file transmitted complies with the format expected   00200
> File not compliant with the format
> Check riga: 91 - colonna: 77  The value '2015-06-04+1:00' of element 'DataDecorrenzaPenale' is not valid.   Check that the file transmitted complies with the format expected   00200
> 
> Quindi direi che il file non è da accettare
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/eLBati/l10n-italy/pull/61/files#r31793002.

## 

Best Regards
Alessio Gerace
-   Think Studio*
   +39 3396911693
   www.thinkstudio.it

Comment (archetipo) [inline l10n_it_fatturapa_in/wizard/wizard_import_fatturapa.py]: corretto il file di test ora passa la verifica SDI

2015-06-05 10:02 GMT+02:00 Alessio Gerace alessio.gerace@gmail.com:

> grazie verifico..
> 
> 2015-06-05 9:41 GMT+02:00 Lorenzo Battistini notifications@github.com:
> 
> > In l10n_it_fatturapa_in/wizard/wizard_import_fatturapa.py
> > https://github.com/eLBati/l10n-italy/pull/61#discussion_r31793002:
> > 
> > > @@ -1274,6 +1299,15 @@ def importFatturaPA(self, cr, uid, ids, context=None):
> > >                  xml_string = fatturapa_attachment.datas.decode('base64')
> > >              xml_string = self.remove_xades_sign(xml_string)
> > >              xml_string = self.strip_xml_content(xml_string)
> > > -            # Fix Date format
> > > -            # Date format must be YYYY-MM-DD
> > > -            # DateType in XSD implements
> > > -            # restrict xs:date minInclusive YYYY-MM-DD  and date should be
> > > -            # in others not parsable ISO 8601:2004 formats like:
> > > -            # YYYY-MM-DDT00:00:00+00:00
> > > -            # YYYY-MM-DD+00:00
> > > -            # YYYY-MM-DDT00:00:00Z
> > 
> > Caricando il file IT05979361218_007 su
> > http://sdi.fatturapa.gov.it/SdI2FatturaPAWeb/AccediAlServizioAction.do?pagina=controlla_fattura
> > ottengo
> > 
> > File not compliant with the format
> > Check riga: 85 - colonna: 89  '2015-06-03T10:21:58+00:00' is not a valid value for 'date'.  Check that the file transmitted complies with the format expected   00200
> > File not compliant with the format
> > Check riga: 85 - colonna: 89  The value '2015-06-03T10:21:58+00:00' of element 'DataScadenzaPagamento' is not valid.    Check that the file transmitted complies with the format expected   00200
> > File not compliant with the format
> > Check riga: 91 - colonna: 77  '2015-06-04+1:00' is not a valid value for 'date'.    Check that the file transmitted complies with the format expected   00200
> > File not compliant with the format
> > Check riga: 91 - colonna: 77  The value '2015-06-04+1:00' of element 'DataDecorrenzaPenale' is not valid.   Check that the file transmitted complies with the format expected   00200
> > 
> > Quindi direi che il file non è da accettare
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/eLBati/l10n-italy/pull/61/files#r31793002.
> 
> ## 
> 
> Best Regards
> Alessio Gerace
> -   Think Studio*
>    +39 3396911693
>    www.thinkstudio.it

## 

Best Regards
Alessio Gerace
-   Think Studio*
   +39 3396911693
   www.thinkstudio.it

**Comment by eLBati** [comment]
Che errore si ottiene senza questa modifica?

**Comment by archetipo** [comment]
SimpleTypeValueError: Type {
http://www.fatturapa.gov.it/sdi/fatturapa/v1.1}DataFatturaType cannot be
created from: (u'2015-04-09+02:00',)

Il giorno 5 giugno 2015 17:15, Lorenzo Battistini notifications@github.com
ha scritto:

> Che errore si ottiene senza questa modifica?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/eLBati/l10n-italy/pull/61#issuecomment-109324426.

## 

Best Regards
Alessio Gerace
-   Think Studio*
   +39 3396911693
   www.thinkstudio.it

**Comment by eLBati** [comment]
Teoricamente, visto che il formato accettato da fatturapa è ISO 8601:2004, pyxb non dovrebbe sollevare eccezione.
Riesci a trovare dove viene configurato in pyxb il formato accettato?

**Comment by archetipo** [comment]
Teoricamente si, ma praticamente lo fa.
riporto un estratto dell'errore:
  File "/usr/local/lib/python2.7/dist-packages/pyxb/binding/basis.py", line 1600, in **call**
    rv = self.typeDefinition().Factory(_args, *_kw)
  File "/usr/local/lib/python2.7/dist-packages/pyxb/binding/basis.py", line 305, in Factory
    rv = cls._DynamicCreate(_args, *_kw)
  File "/usr/local/lib/python2.7/dist-packages/pyxb/binding/basis.py", line 679, in _DynamicCreate
    raise pyxb.SimpleTypeValueError(ctor, args)
SimpleTypeValueError: Type {http://www.fatturapa.gov.it/sdi/fatturapa/v1.1}DataFatturaType cannot be created from: (u'2015-04-09+02:00',)

quindi credo si debba indagare in basis.py nel core di pyxb per verificare cio' che accade.

**Comment by eLBati** [comment]
Appunto dico: visto che pyxb non dovrebbe sollevare eccezione ma invece lo fa, la soluzione corretta sarebbe modificare il binding in modo da fargli accettare i formati che deve accettare.
Secondo te riesci a trovarlo o è troppo oneroso?

**Comment by archetipo** [comment]
posso fare una analisi preliminare ed approfondire un po la ricerca e
capire quanto effettivamente puo' essere oneroso.. perche' a priori non
saprei proprio dire, dato  che non ho idea di come sia progettato PyXB.

Il giorno 8 giugno 2015 12:33, Lorenzo Battistini notifications@github.com
ha scritto:

> Appunto dico: visto che pyxb non dovrebbe sollevare eccezione ma invece lo
> fa, la soluzione corretta sarebbe modificare il binding in modo da fargli
> accettare i formati che deve accettare.
> Secondo te riesci a trovarlo o è troppo oneroso?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/eLBati/l10n-italy/pull/61#issuecomment-109944972.

## 

Best Regards
Alessio Gerace
-   Think Studio*
   +39 3396911693
   www.thinkstudio.it

**Comment by archetipo** [comment]
Ho provato a dare un occhio a capire un po' come lavora il pyxb, riprovando
a generarne nuovamente i bindings, ed ho scoperto che il problema sembra
risiedere nello schema, mi spiego meglio:
lo schema FatturaV1.1.xsd definisce il datetype:
-  <xs:simpleType name="DataFatturaType">     <xs:restriction
  base="xs:date">       <xs:minInclusive value="1970-01-01" />
  /xs:restriction   /xs:simpleType*

il tag:

*<xs:minInclusive value="1970-01-01" /> *

riporta il formato minimo del valore che puo' contenere il campo, che
eredita da xs,date http://www.w3schools.com/schema/schema_dtypes_date.asp ,
che pero' in realta' e' superfluo perche' sottointeso da xs.date,
e quindi al massimo doveva essere riportato come attriuto maxInclusive.

questa specificazione incompleta nel binding fatturapa_v_1_1.py viene
tradotto in:

_DataFatturaType._CF___minInclusive = pyxb.binding.facets.CF__
_minInclusive(    value_datatype=_
_DataFatturaType,    value=pyxb.binding.datatypes._

_date('1970-01-01'))DataFatturaType._InitializeFacetMap(DataFatturaType._CF_minInclusive)_
in sostanza la regola gia' citata in precedenza che indica il formato
minimo dei valori del campo.
questo pero' poi viene applicato all' inizializzazione dell oggetto e
quindi genera l'errore.
Nel mio test ho provato a modificare lo schema in questo modo:
-  <xs:simpleType name="DataFatturaType">     <xs:restriction
  base="xs:date">     /xs:restriction   /xs:simpleType*

a generare i nuovi bindings e copiare la classe DataFatturaType che questa
volta era dichiarata semplicemente:

_DataFatturaType._InitializeFacetMap()_

Questa modifica , basterebbe a risolvere il problema , ma poi nel caso al
prossimo versione dello schema e quindi alla rigenerazione dei nuovi
bindings , se non rivedranno *DataFatturaType *ci ritroveremo nuovamente
l'errore.

Il giorno 8 giugno 2015 12:45, Alessio Gerace alessio.gerace@gmail.com ha
scritto:

> posso fare una analisi preliminare ed approfondire un po la ricerca e
> capire quanto effettivamente puo' essere oneroso.. perche' a priori non
> saprei proprio dire, dato  che non ho idea di come sia progettato PyXB.
> 
> Il giorno 8 giugno 2015 12:33, Lorenzo Battistini <
> notifications@github.com> ha scritto:
> 
> > Appunto dico: visto che pyxb non dovrebbe sollevare eccezione ma invece
> > lo fa, la soluzione corretta sarebbe modificare il binding in modo da
> > fargli accettare i formati che deve accettare.
> > Secondo te riesci a trovarlo o è troppo oneroso?
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/eLBati/l10n-italy/pull/61#issuecomment-109944972.
> 
> ## 
> 
> Best Regards
> Alessio Gerace
> -   Think Studio*
>    +39 3396911693
>    www.thinkstudio.it

## 

Best Regards
Alessio Gerace
-   Think Studio*
   +39 3396911693
   www.thinkstudio.it

**Comment by eLBati** [comment]
Grazie alle tue indagini ho fixato con
https://github.com/archetipo/l10n-italy/commit/8f0fcc00c54ac1f783a13d8e9c1a0d9139f7070c

Se sei d'accordo con le modifiche, mergio

**Pull request merged**